src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1916 @@
+        {
+            foreach (var child in node)
+            {
+                if (child.IsNode)
+                {
+                    CollectDiagnostics(child.Node, diagnostics);
+                }
+                else
+                {
+                    var token = child.Token;
+                    foreach (var trivia in token.LeadingTrivia)
+                    {
+                        AddUniqueDiagnostics(trivia.Diagnostics, diagnostics);
+                    }
+
+                    AddUniqueDiagnostics(token.Diagnostics, diagnostics);
+                }
+            }
+        }
+
+        /// <summary>
+        /// It's very common to have duplicated diagnostics.  For example, consider "((". This will
+        /// have two 'missing )' diagnostics, both at the end.  Reporting both isn't helpful, so we
+        /// filter duplicates out here.
+        /// </summary>
+        private static void AddUniqueDiagnostics(ImmutableArray<EmbeddedDiagnostic> from, ArrayBuilder<EmbeddedDiagnostic> to)
+        {
+            foreach (var diagnostic in from)
+            {
+                if (!to.Contains(diagnostic))
+                {
+                    to.Add(diagnostic);
+                }
+            }
+        }
+
+        private RegexExpressionNode ParseAlternatingSequences(bool consumeCloseParen)
+        {
+            try
+            {
+                _recursionDepth++;
+                StackGuard.EnsureSufficientExecutionStack(_recursionDepth);
+                return ParseAlternatingSequencesWorker(consumeCloseParen);
+            }
+            finally
+            {
+                _recursionDepth--;
+            }
+        }
+
+        /// <summary>
+        /// Parses out code of the form: ...|...|...
+        /// This is the type of code you have at the top level of a regex, or inside any grouping
+        /// contruct.  Note that sequences can be empty in .net regex.  i.e. the following is legal:
+        /// 
+        ///     ...||...
+        /// 
+        /// An empty sequence just means "match at every position in the test string".
+        /// </summary>
+        private RegexExpressionNode ParseAlternatingSequencesWorker(bool consumeCloseParen)
+        {
+            RegexExpressionNode current = ParseSequence(consumeCloseParen);
+
+            while (_currentToken.Kind == RegexKind.BarToken)
+            {
+                // Trivia allowed between the | and the next token.
+                current = new RegexAlternationNode(
+                    current, ConsumeCurrentToken(allowTrivia: true), ParseSequence(consumeCloseParen));
+            }
+
+            return current;
+        }
+
+        private RegexSequenceNode ParseSequence(bool consumeCloseParen)
+        {
+            var list = ArrayBuilder<RegexExpressionNode>.GetInstance();
+
+            if (ShouldConsumeSequenceElement(consumeCloseParen))
+            {
+                do
```

Comment:
Could this be a `while` loop instead?

Reply:
not sure if that helps.  perhaps you could demonstrate how this would work?

Reply:
Here's what I meant:

```C#
                if (!ShouldConsumeSequenceElement(consumeCloseParen))
                {
                    return ImmutableArray<RegexExpressionNode>.Empty;
                }

                do
                {
                    ...
                }
                while (ShouldConsumeSequenceElement(consumeCloseParen));

               return new RegexSequenceNode(list.ToImmutableAndFree());
```

or the simpler one without early bail-out:

```C#
              while (ShouldConsumeSequenceElement(consumeCloseParen));
              {
                  ...
              }

               return new RegexSequenceNode(list.ToImmutableAndFree());
```

---
In reply to: [178469926](https://github.com/dotnet/roslyn/pull/23984#discussion_r178469926) [](ancestors = 178469926)

Reply:
Yes... that does seem better.  I wonder why i wrote it that way :)